File: chapter 2/learning_service/scripts/turtle_spawn.py
#!/usr/bin/env python
# coding:utf-8

import sys

from requests import request
import rospy
from turtlesim.srv import Spawn

def turtle_spawn():
	# ROS节点初始化
    rospy.init_node('turtle_spawn')

	# 发现/spawn服务后，创建一个服务客户端，连接名为/spawn的service
    rospy.wait_for_service('/spawn')
    try:
        spawn_turtle = rospy.ServiceProxy('/spawn', Spawn)

        srv = Spawn()
        srv.x     = 1.0;
        srv.y     = 1.0;
        srv.theta = 0.0;
        srv.name  = "turtle2";
        
        response = spawn_turtle(srv.x, srv.y, srv.theta, srv.name)

        print("Call service to spawn turtle [x:%0.6f, y:%0.6f, name:%s]" %(srv.x,srv.y,srv.name))
        print("Spawn turtle successfully, new turtle name:%s" %response.name)
    except rospy.ServiceException as e:
        print ("Service call failed: %s"%e)

if __name__ == "__main__":
	turtle_spawn()
   
    

File: chapter 2/learning_service/src/turtle_spawn.cpp
#include <ros/ros.h>
#include <turtlesim/Spawn.h>

int main(int argc, char** argv)
{
	ros::init(argc, argv, "turtle_spawn");

	ros::NodeHandle nh;

	ros::service::waitForService("/spawn");
	ros::ServiceClient spawn_turtle = nh.serviceClient<turtlesim::Spawn>("/spawn");

	turtlesim::Spawn srv;
	srv.request.x = 1.0;
	srv.request.y = 1.0;
	srv.request.name = "turtle2";

	ROS_INFO("Call service to spwan turtle [x:%0.6f, y:%0.6f, name:%s]", 
			 srv.request.x, srv.request.y, srv.request.name.c_str());

	spawn_turtle.call(srv); 

	ROS_INFO("Spwan turtle successfully, new turtle name:%s", srv.response.name.c_str());

	return 0;
};
